core/factorization/par_ilu.cpp

```diff
@@ -93,14 +100,34 @@ ParIlu<ValueType, IndexType>::generate_l_u(
     }
 
     const auto matrix_size = csr_system_matrix->get_size();
-    size_type l_nnz{};
-    size_type u_nnz{};
-    exec->run(par_ilu_factorization::make_compute_nnz_l_u(csr_system_matrix,
-                                                          &l_nnz, &u_nnz));
-    auto l_factor =
-        l_matrix_type::create(exec, matrix_size, l_nnz /* TODO set strategy */);
-    auto u_factor =
-        u_matrix_type::create(exec, matrix_size, u_nnz /* TODO set strategy */);
+    const auto number_rows = matrix_size[0];
+    Array<IndexType> l_row_ptrs{exec, number_rows + 1};
+    Array<IndexType> u_row_ptrs{exec, number_rows + 1};
+    exec->run(par_ilu_factorization::make_initialize_row_ptrs_l_u(
+        csr_system_matrix, l_row_ptrs.get_data(), u_row_ptrs.get_data()));
+
+    IndexType l_nnz_it;
+    IndexType u_nnz_it;
+    // Since nnz is always at row_ptrs[m], it can be extracted easily
+    host_exec->copy_from(exec.get(), 1, l_row_ptrs.get_data() + number_rows,
+                         &l_nnz_it);
+    host_exec->copy_from(exec.get(), 1, u_row_ptrs.get_data() + number_rows,
+                         &u_nnz_it);
+    auto l_nnz = static_cast<size_type>(l_nnz_it);
+    auto u_nnz = static_cast<size_type>(u_nnz_it);
+
+    // Since `row_ptrs` of L and U is already created, the matrix can be
+    // directly created with it
+    Array<IndexType> l_col_idxs{exec, l_nnz};
+    Array<ValueType> l_vals{exec, l_nnz};
+    auto l_factor = l_matrix_type::create(
+        exec, matrix_size, std::move(l_vals), std::move(l_col_idxs),
+        std::move(l_row_ptrs) /* TODO set strategy */);
```

Comment:
/* TODO set strategy */ ?

Reply:
I don't know what Csr strategy to use, so I put it as a `TODO`. In general, I am not sure when which strategy is best, so I left it as the default one.
Should we put the strategy as a parameter we can set (like the number of iterations)?

Reply:
Here is your answer. The automatic strategy is quite good.

![Ginkgo CSR](https://i.postimg.cc/tRqMPLXt/v100-performance-all-csr.png)

Reply:
Thanks for the information. However, since we need a `CudaExecutor` for the `Csr::automatical` strategy, I decided to stay with the default case, which is `Csr::cusparse`.
I will create an Issue about this and the general usage of the current strategies.